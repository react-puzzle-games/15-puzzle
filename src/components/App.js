// @ts-check

import PropTypes from "prop-types";
import React, { Component } from "react";
import levelFactory from "./../lib/levels-factory";
import Footer from "./Footer";
import Game from "./Game";

class App extends Component {
  constructor(props) {
    super(props);

    const level = props.level ? props.level : levelFactory(4 ** 2);
    const originalLevel = Object.assign({}, level);

    this.state = {
      original: originalLevel,
      level,
    };
  }

  onResetClick = () => {
    this.setState({
      level: {
        tileSet: this.state.original.tileSet,
      },
    });
  };

  onNewClick = () => {
    const newLevel = levelFactory(4 ** 2);
    const newOriginalLevel = Object.assign({}, newLevel);
    this.setState({
      level: newLevel,
      original: newOriginalLevel,
    });
  };

  render() {
    return (
      <main
        style={{
          display: "flex",
          flexDirection: "column",
          minHeight: "100vh",
        }}
      >
        <Game
          gridSize={4}
          tileSize={90}
          numbers={this.state.level.tileSet}
          onResetClick={this.onResetClick}
          onNewClick={this.onNewClick}
          original={this.state.original.tileSet}
        />
        <Footer />
      </main>
    );
  }
}

App.propTypes = {
  level: PropTypes.shape({
    tileSet: PropTypes.arrayOf(PropTypes.number).isRequired,
  }),
};

export default App;
