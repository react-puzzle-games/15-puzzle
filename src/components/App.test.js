// @ts-check

import React from "react";
import renderer from "react-test-renderer";
import App from "./App";

test("App renders correctly", () => {
  const tree = renderer
    .create(
      <App
        level={{
          tileSet: [4, 6, 11, 1, 2, 16, 3, 15, 10, 9, 7, 13, 12, 14, 8, 5],
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
