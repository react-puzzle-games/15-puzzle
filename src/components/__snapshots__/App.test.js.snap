// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<main
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "minHeight": "100vh",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
      }
    }
  >
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-root-1 MuiPaper-elevation4"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          className="MuiTypography-root makeStyles-title-2 MuiTypography-h6"
        >
          React Puzzle Games - 15 Puzzle
        </div>
        <button
          aria-label="Start a new game"
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root menuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M13 3h-2v10h2V3zm4.83 2.17l-1.42 1.42C17.99 7.86 19 9.81 19 12c0 3.87-3.13 7-7 7s-7-3.13-7-7c0-2.19 1.01-4.14 2.58-5.42L6.17 5.17C4.23 6.82 3 9.26 3 12c0 4.97 4.03 9 9 9s9-4.03 9-9c0-2.74-1.23-5.18-3.17-6.83z"
                />
              </svg>
            </span>
            <span
              className="MuiTypography-root MuiTypography-button"
            >
              New game
            </span>
          </span>
        </button>
        <button
          aria-label="Pause/Continue current game."
          className="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root menuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
                />
              </svg>
            </span>
            <span
              className="MuiTypography-root MuiTypography-button"
            >
              Pause
            </span>
          </span>
        </button>
        <button
          aria-label="Reset game"
          className="MuiButtonBase-root MuiButton-root MuiButton-text"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            <span
              className="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                />
              </svg>
            </span>
            Reset game
          </span>
        </button>
        <div
          className="MuiChip-root"
          onKeyDown={[Function]}
          onKeyUp={[Function]}
        >
          <div
            className="MuiAvatar-root MuiAvatar-circular MuiChip-avatar MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 5.72l-4.6-3.86-1.29 1.53 4.6 3.86L22 5.72zM7.88 3.39L6.6 1.86 2 5.71l1.29 1.53 4.59-3.85zM12.5 8H11v6l4.75 2.85.75-1.23-4-2.37V8zM12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9c4.97 0 9-4.03 9-9s-4.03-9-9-9zm0 16c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
              />
            </svg>
          </div>
          <span
            className="MuiChip-label"
          >
            <span
              className="MuiTypography-root MuiTypography-body1"
            >
              0
              s
            </span>
          </span>
        </div>
        <div
          className="MuiChip-root"
          onKeyDown={[Function]}
          onKeyUp={[Function]}
        >
          <div
            className="MuiAvatar-root MuiAvatar-circular MuiChip-avatar MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
              />
            </svg>
          </div>
          <span
            className="MuiChip-label"
          >
            <span
              className="MuiTypography-root MuiTypography-body1"
            >
              0
            </span>
          </span>
        </div>
      </div>
    </header>
    <div
      className="makeStyles-root-3 makeStyles-root-5"
    >
      <div
        className="makeStyles-tile-4 makeStyles-tile-6"
      >
        <div
          className="makeStyles-root-7 makeStyles-root-9"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            4
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-10"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            6
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-11"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            11
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-12"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            1
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-13"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            2
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-14"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            16
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-15"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            3
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-16"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            15
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-17"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            10
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-18"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            9
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-19"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            7
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-20"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            13
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-21"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            12
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-22"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            14
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-23"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            8
          </span>
        </div>
        <div
          className="makeStyles-root-7 makeStyles-root-24"
          onClick={[Function]}
        >
          <span
            className="makeStyles-tileNumber-8"
          >
            5
          </span>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="makeStyles-root-25"
  >
    <a
      href="https://github.com/react-puzzle-games/15-puzzle/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        className="makeStyles-root-28"
        height="24px"
        version="1.1"
        viewBox="0 0 435.549 435.549"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M409.132,114.573c-19.608-33.596-46.205-60.194-79.798-79.8C295.736,15.166,259.057,5.365,219.271,5.365 c-39.781,0-76.472,9.804-110.063,29.408c-33.596,19.605-60.192,46.204-79.8,79.8C9.803,148.168,0,184.854,0,224.63 c0,47.78,13.94,90.745,41.827,128.906c27.884,38.164,63.906,64.572,108.063,79.227c5.14,0.954,8.945,0.283,11.419-1.996 c2.475-2.282,3.711-5.14,3.711-8.562c0-0.571-0.049-5.708-0.144-15.417c-0.098-9.709-0.144-18.179-0.144-25.406l-6.567,1.136 c-4.187,0.767-9.469,1.092-15.846,1c-6.374-0.089-12.991-0.757-19.842-1.999c-6.854-1.231-13.229-4.086-19.13-8.559 c-5.898-4.473-10.085-10.328-12.56-17.556l-2.855-6.57c-1.903-4.374-4.899-9.233-8.992-14.559 c-4.093-5.331-8.232-8.945-12.419-10.848l-1.999-1.431c-1.332-0.951-2.568-2.098-3.711-3.429c-1.142-1.331-1.997-2.663-2.568-3.997 c-0.572-1.335-0.098-2.43,1.427-3.289c1.525-0.859,4.281-1.276,8.28-1.276l5.708,0.853c3.807,0.763,8.516,3.042,14.133,6.851 c5.614,3.806,10.229,8.754,13.846,14.842c4.38,7.806,9.657,13.754,15.846,17.847c6.184,4.093,12.419,6.136,18.699,6.136 c6.28,0,11.704-0.476,16.274-1.423c4.565-0.952,8.848-2.383,12.847-4.285c1.713-12.758,6.377-22.559,13.988-29.41 c-10.848-1.14-20.601-2.857-29.264-5.14c-8.658-2.286-17.605-5.996-26.835-11.14c-9.235-5.137-16.896-11.516-22.985-19.126 c-6.09-7.614-11.088-17.61-14.987-29.979c-3.901-12.374-5.852-26.648-5.852-42.826c0-23.035,7.52-42.637,22.557-58.817 c-7.044-17.318-6.379-36.732,1.997-58.24c5.52-1.715,13.706-0.428,24.554,3.853c10.85,4.283,18.794,7.952,23.84,10.994 c5.046,3.041,9.089,5.618,12.135,7.708c17.705-4.947,35.976-7.421,54.818-7.421s37.117,2.474,54.823,7.421l10.849-6.849 c7.419-4.57,16.18-8.758,26.262-12.565c10.088-3.805,17.802-4.853,23.134-3.138c8.562,21.509,9.325,40.922,2.279,58.24 c15.036,16.18,22.559,35.787,22.559,58.817c0,16.178-1.958,30.497-5.853,42.966c-3.9,12.471-8.941,22.457-15.125,29.979 c-6.191,7.521-13.901,13.85-23.131,18.986c-9.232,5.14-18.182,8.85-26.84,11.136c-8.662,2.286-18.415,4.004-29.263,5.146 c9.894,8.562,14.842,22.077,14.842,40.539v60.237c0,3.422,1.19,6.279,3.572,8.562c2.379,2.279,6.136,2.95,11.276,1.995 c44.163-14.653,80.185-41.062,108.068-79.226c27.88-38.161,41.825-81.126,41.825-128.906 C438.536,184.851,428.728,148.168,409.132,114.573z"
          />
        </g>
      </svg>
    </a>
    <p>
      Made with 
      <span
        className="makeStyles-heart-26"
      >
        ♥
      </span>
       by
       
      <a
        className="makeStyles-link-27"
        href="https://github.com/react-puzzle-games"
        rel="noopener noreferrer"
        target="_blank"
      >
         
        React Puzzle Games
      </a>
    </p>
  </footer>
</main>
`;
